src/assets/src/hooks/useEntityChanges.ts

```diff
@@ -0,0 +1,27 @@
+import { useState } from "react";
+
+import { ChangeEvent, compareEntities, ComparableEntity } from "../changes";
+
+
+export function useEntityChanges<T extends ComparableEntity>():
+    [ChangeEvent[], (oldEntities: readonly T[], newEntities: readonly T[]) => void, (key: number) => void]
+{
+    const [changeEvents, setChangeEvents] = useState([] as ChangeEvent[]);
+    const [nextID, setNextID] = useState(0);
+
+    const compareAndSetEventMap = (oldEntities: readonly T[], newEntities: readonly T[]): void => {
+        const changeText = compareEntities<T>(oldEntities.slice(), newEntities.slice());
+        if (changeText !== undefined) {
+            const newChangeEvent = { eventID: nextID, text: changeText } as ChangeEvent;
+            setChangeEvents([...changeEvents, newChangeEvent]);
+            setNextID(nextID + 1);
+        }
+    };
+
+    // https://reactjs.org/docs/hooks-reference.html#functional-updates
+    const deleteChangeEvent = (id: number) => {
+        setChangeEvents((prevChangeEvents) => prevChangeEvents.filter((e) => id !== e.eventID));
+    };
```

Comment:
What would you think about giving each event a `timestamp` and filtering everything >= `n` seconds old, instead of making events self-destruct? Seems like it could reduce some state, although conceivably the list could eventually grow pretty large...

Reply:
Can you elaborate on how you think this would "reduce some state"? I would think that storing less data whenever possible would be preferable, though I'm sure there's a performance cost to all the updates (though I haven't seen this being a problem).

Reply:
I mean it could remove the `useEffect` from `TimedChangeAlert` as well as the need for the deletion callback. Maybe "state" isn't the right word for it, but rather "side effect," although react hooks kinda muddle the difference. At any rate, no worries & this is good; now that I think about it, maybe the alternative of checking the timestamp against `now` on each render cycle has hidden complications.